metrics/user_registries.go

```diff
@@ -0,0 +1,809 @@
+		level.Warn(r.logger).Log("msg", "failed to gather metrics from registry", "user", ur.user, "err", err)
+		return false
+	}
+
+	for ix := 0; ix < len(last); {
+		// Only keep metrics for which we don't want to go down, since that indicates reset (counter, summary, histogram).
+		switch last[ix].GetType() {
+		case dto.MetricType_COUNTER, dto.MetricType_SUMMARY, dto.MetricType_HISTOGRAM:
+			ix++
+		default:
+			// Remove gauges and unknowns.
+			last = append(last[:ix], last[ix+1:]...)
+		}
+	}
+
+	// No metrics left.
+	if len(last) == 0 {
+		return false
+	}
+
+	ur.lastGather, err = NewMetricFamilyMap(last)
+	if err != nil {
+		level.Warn(r.logger).Log("msg", "failed to gather metrics from registry", "user", ur.user, "err", err)
+		return false
+	}
+
+	ur.user = ""
+	ur.reg = nil
+	return true
+}
+
+// Registries returns a copy of the user registries list.
+func (r *UserRegistries) Registries() []UserRegistry {
+	r.regsMu.Lock()
+	defer r.regsMu.Unlock()
+
+	out := make([]UserRegistry, 0, len(r.regs))
+	out = append(out, r.regs...)
+
+	return out
+}
+
+// GetRegistryForUser returns currently active registry for given user, or nil, if there is no such registry.
+func (r *UserRegistries) GetRegistryForUser(user string) *prometheus.Registry {
+	r.regsMu.Lock()
+	defer r.regsMu.Unlock()
+
+	for idx := 0; idx < len(r.regs); idx++ {
+		if user != r.regs[idx].user {
+			continue
+		}
+
+		return r.regs[idx].reg
+	}
+	return nil
+}
+
+func (r *UserRegistries) BuildMetricFamiliesPerUser() MetricFamiliesPerUser {
+	data := MetricFamiliesPerUser{}
+	for _, entry := range r.Registries() {
+		// Set for removed users.
+		if entry.reg == nil {
+			if entry.lastGather != nil {
+				data = append(data, struct {
+					user    string
+					metrics MetricFamilyMap
+				}{user: "", metrics: entry.lastGather})
+			}
+
+			continue
+		}
+
+		m, err := entry.reg.Gather()
+		if err == nil {
+			var mfm MetricFamilyMap // := would shadow err from outer block, and single err check will not work
+			mfm, err = NewMetricFamilyMap(m)
+			if err == nil {
+				data = append(data, struct {
+					user    string
+					metrics MetricFamilyMap
+				}{
+					user:    entry.user,
+					metrics: mfm,
+				})
+			}
+		}
+
+		if err != nil {
+			level.Warn(r.logger).Log("msg", "failed to gather metrics from registry", "user", entry.user, "err", err)
+			continue
+		}
+	}
+	return data
+}
+
+// FromLabelPairsToLabels converts dto.LabelPair into labels.Labels.
+func FromLabelPairsToLabels(pairs []*dto.LabelPair) labels.Labels {
+	builder := labels.NewBuilder(nil)
+	for _, pair := range pairs {
+		builder.Set(pair.GetName(), pair.GetValue())
+	}
+	return builder.Labels(nil)
+}
+
+// GetSumOfHistogramSampleCount returns the sum of samples count of histograms matching the provided metric name
+// and optional label matchers. Returns 0 if no metric matches.
+func GetSumOfHistogramSampleCount(families []*dto.MetricFamily, metricName string, matchers labels.Selector) uint64 {
```

Comment:
There's a problem with this function.

For context, it is only used in these two places https://github.com/grafana/mimir/blob/eb812fd67b21e89b055e88946e03b321455551bc/pkg/distributor/ha_tracker_test.go#L255-L263

But it imports the whole of Prometheus because it needs the `labels` package to use `matchers`, I took a look at https://github.com/prometheus/common, and there's no such struct (that allows us to do regex matching on labels).

The way I see it we have a couple of options:

- Import Prometheus as part of dskit (there's no precedent for this it seems)
- Take `common.Matcher` from the silence package in prometheus/common and attempt to do the comparison that way. 
- Rewrite the test that uses this function in Mimir

Reply:
I've removed it for now - think it's the safest option.